### PR TITLE
Remove `skipOnError` from `StopOnError` and simplify it code

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  |

### CODE REVIEW

**Comment by arogachev** [comment]
Synced with master changes, resolved merge conflicts. Seems like something is wrong with Github diff again. Locally it's OK. The only changed files:

- `src/Rule/StopOnError.php`
- `src/Rule/StopOnErrorHandler.php`
- `src/ValidationContext.php`
- `src/Validator.php`
- `tests/Rule/StopOnErrorTest.php`

File: src/Rule/StopOnError.php [hunk 1]
```diff
@@ -19,37 +18,54 @@
...
+ * $rule = new StopOnError([
+ *      new HasLength(min: 3),
+ *      // This operation executes DB query and thus heavier. It's preferable not to call it if the previous rule did
+ *      not pass the validation.
+ *      new ExistsInDatabase(),
+ * ]);
+ * ```
+ *
+ * Not to be confused with skipping, there is a separate functionality for that, see {@see SkipOnErrorInterface}.
+ *
```
Comment (arogachev) [inline src/Rule/StopOnError.php:35]:
```suggestion
 *
 * ```php
 * $rules = [
 *      new SimpleRule1(), // Let's say there is an error.
 *      // Then this rule is skipped completely with all its related rules.
 *      new StopOnError([
 *          new HeavyRule1(), // Skipped.
 *          new HeavyRule2(), // Skipped.
 *     ]),
 *     new SimpleRule2(), // Skipping of intermediate rules depends on `skipOnError` option.
 * ]);
 * ```
 *
 * Use grouping / ordering / `skipOnError` option to achieve the desired effect,
 *
```

**Review by arogachev — CHANGES_REQUESTED** [review]
See suggestions.

File: src/Validator.php [hunk 2]
```diff
@@ -183,7 +177,7 @@ private function validateInternal(mixed $value, iterable $rules, ValidationConte
                 continue;
             }
 
-            $context->setParameter(self::PARAMETER_PREVIOUS_RULES_ERRORED, true);
+            $context->setParameter(ValidationContext::PARAMETER_PREVIOUS_RULES_ERRORED, true);
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Validator.php:180]: Why?

File: src/ValidationContext.php [hunk 3]
```diff
@@ -14,6 +15,13 @@
  */
 final class ValidationContext
 {
+    /**
+     * @const A name of parameter indicating that previous rule in the set caused validation error. Used in
+     * {@see Validator} with {@see SkipOnErrorInterface} to allow skipping of the current rule if its configuration
+     * allows it. Used in {@see StopOnError} rule also.
+     */
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/ValidationContext.php]: done

File: tests/Rule/StopOnErrorTest.php [hunk 4]
```diff
@@ -113,17 +111,20 @@ public function dataValidationFailed(): array
...
                 ['' => ['This value must contain at most 1 character.']],
             ],
+            'case3' => [
+                'hello',
+                [
+                    new Number(),
+                    new StopOnError([
+                        new HasLength(max: 1),
+                        new HasLength(min: 10),
+                    ]),
```
[the remark this answers is not in the document]
Comment (vjik) [reply tests/Rule/StopOnErrorTest.php:121]: done

**Review by arogachev — APPROVED** [review]

**Pull request merged**